staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go

```diff
@@ -75,6 +76,46 @@ func ValidateStructural(s *Structural, fldPath *field.Path) field.ErrorList {
 	return allErrs
 }
 
+// ValidateStructuralUpdate checks invariants of a structural schema on update or creation.
+// in publishing, we make old being nil to keep backwards compatibility.
+// TODO: reuse the function in v1 validation
+func ValidateStructuralUpdate(s *Structural, old *Structural, fldPath *field.Path) field.ErrorList {
+	if s == nil {
+		return nil
+	}
+
+	allErrs := field.ErrorList{}
+
+	if old != nil {
```

Comment:
Ah, if the scope of this pr is only in openapi publishing, we can drop the `old *Structural` bit to make it more concrete and less confusing. Given that traversing JSONSchemaProps seems to more complex than traversing Structural schema, I'm not sure how re-usable this code will be in v1. Otherwise LGTM

I'll let @liggitt to make the final call

Reply:
let's limit this to publishing. I'll tighten validation for v1 CRDs in https://github.com/kubernetes/kubernetes/pull/79604